### ISSUE TITLE
implement regression from scratch

### ISSUE THREAD

**Issue by almsam** [almsam/Numerical-Analysis-code#56]
the issue with the project as of now is the regression from statsmodels is outputting objects not neccesarily of a type we would prefer. The issue is to implement multiple linear regression from scratch to ensure object oriented control of the output models to ensure regression works better

**Comment by Ranakghosh7** [comment]
import numpy as np

class MyLinearRegression:
    def __init__(self, learning_rate=0.01, epochs=1000):
        self.learning_rate = learning_rate
        self.epochs = epochs
        self.weights = None
        self.bias = None

    def fit(self, X, y):
        """Train the model using gradient descent"""
        n_samples, n_features = X.shape
        self.weights = np.zeros(n_features)
        self.bias = 0

        for _ in range(self.epochs):
            y_pred = np.dot(X, self.weights) + self.bias
            error = y_pred - y

            # Compute gradients
            dw = (1 / n_samples) * np.dot(X.T, error)
            db = (1 / n_samples) * np.sum(error)

            # Update weights
            self.weights -= self.learning_rate * dw
            self.bias -= self.learning_rate * db

    def predict(self, X):
        """Predict house prices"""
        return np.dot(X, self.weights) + self.bias

    def coefficients(self):
        """Return model parameters"""
        return self.weights, self.bias

**Comment by almsam** [comment]
> import numpy as np
> 
> class MyLinearRegression: def **init**(self, learning_rate=0.01, epochs=1000): self.learning_rate = learning_rate self.epochs = epochs self.weights = None self.bias = None
> 
> ```
> def fit(self, X, y):
>     """Train the model using gradient descent"""
>     n_samples, n_features = X.shape
>     self.weights = np.zeros(n_features)
>     self.bias = 0
> 
>     for _ in range(self.epochs):
>         y_pred = np.dot(X, self.weights) + self.bias
>         error = y_pred - y
> 
>         # Compute gradients
>         dw = (1 / n_samples) * np.dot(X.T, error)
>         db = (1 / n_samples) * np.sum(error)
> 
>         # Update weights
>         self.weights -= self.learning_rate * dw
>         self.bias -= self.learning_rate * db
> 
> def predict(self, X):
>     """Predict house prices"""
>     return np.dot(X, self.weights) + self.bias
> 
> def coefficients(self):
>     """Return model parameters"""
>     return self.weights, self.bias
> ```

a couple issues with this

first of all this implements lin reg using gradient decent - are we sure this is the best option here? consider the method described in [RegressionBasics](https://github.com/almsam/Numerical-Analysis-code/blob/main/RegressionBasics.pdf)

furthermore - you managed to get this done in 5 mins, made comments like those, & made a structure like that - so forgive me but this frankly smells like AI code to me - if thats the case I would prefer if you cited that & gave the prompt to ensure honesty in our work

but besides that - if you want to proceed with this issue as above & contribute here id encorage you to implement lin reg using the sum of squared deviations method ( explained in [RegressionBasics](https://github.com/almsam/Numerical-Analysis-code/blob/main/RegressionBasics.pdf) ) & throw it into a new branch & make a PR

**Comment by Ranakghosh7** [comment]
import numpy as np
import unittest

class LinearRegressionOLS:
    def __init__(self):
        self.weights = None
        self.bias = None

    def fit(self, X, y):
        """Train the model using the sum of squared deviations (OLS normal equation)"""
        n_samples, n_features = X.shape

        # Add bias term (column of ones)
        X_b = np.c_[np.ones((n_samples, 1)), X]

        # Compute optimal parameters using (X^T X)^-1 X^T y
        theta = np.linalg.inv(X_b.T @ X_b) @ X_b.T @ y

        # Extract bias and weights
        self.bias = theta[0]
        self.weights = theta[1:]

    def predict(self, X):
        """Predict house prices using the OLS model"""
        return np.dot(X, self.weights) + self.bias

    def coefficients(self):
        """Return model parameters"""
        return self.weights, self.bias

**Comment by almsam** [comment]
this bit looks good - throw your work into a new branch & make a PR to request a merge

**Comment by Ranakghosh7** [comment]
git checkout -b new-feature-branch
git add .
git commit -m "Refactored OLS model with improved stability and encapsulation"
git push origin new-feature-branch
gh pr create --base main --head new-feature-branch --title "Refactored OLS Model" --body "Improved numerical stability and encapsulated in a class."

I hope this will work

**Comment by almsam** [comment]
#72 estimated to be final
